RecoJets/JetProducers/plugins/FixedGridRhoProducerFastjetFromRecHit.cc

```diff
@@ -0,0 +1,168 @@
+  //PF recHit collection is regional, while the full raw recHit collection is available.
+  desc.add<edm::InputTag>("hbheRecHitsTag", edm::InputTag("hltHbhereco"));
+  desc.add<edm::InputTag>("ebRecHitsTag", edm::InputTag("hltEcalRecHit", "EcalRecHitsEB"));
+  desc.add<edm::InputTag>("eeRecHitsTag", edm::InputTag("hltEcalRecHit", "EcalRecHitsEE"));
+  desc.add<bool>("skipHCAL", false);
+  desc.add<bool>("skipECAL", false);
+  //eThresHB/HE are from RecoParticleFlow/PFClusterProducer/python/particleFlowRecHitHBHE_cfi.py
+  desc.add<std::vector<double> >("eThresHB", {0.1, 0.2, 0.3, 0.3});
+  desc.add<std::vector<double> >("eThresHE", {0.1, 0.2, 0.2, 0.2, 0.2, 0.2, 0.2});
+  desc.add<double>("maxRapidity", 2.5);
+  desc.add<double>("gridSpacing", 0.55);
+  descriptions.addWithDefaultLabel(desc);
+}
+
+FixedGridRhoProducerFastjetFromRecHit::~FixedGridRhoProducerFastjetFromRecHit() = default;
+
+void FixedGridRhoProducerFastjetFromRecHit::produce(edm::Event &iEvent, const edm::EventSetup &iSetup) {
+  std::vector<fastjet::PseudoJet> inputs;
+  auto const &thresholds = iSetup.getData(ecalPFRecHitThresholdsToken_);
+  auto const &caloGeometry = iSetup.getData(caloGeometryToken_);
+
+  if (!skipHCAL_) {
+    auto const &hbheRecHits = iEvent.get(hbheRecHitsTag_);
+    inputs.reserve(inputs.size() + hbheRecHits.size());
+    for (const auto &hit : hbheRecHits) {
+      if (passedHcalNoiseCut(hit)) {
+        const auto &hitp4 = getHitP4(hit.id(), hit.energy(), caloGeometry);
+        inputs.emplace_back(fastjet::PseudoJet(hitp4[0], hitp4[1], hitp4[2], hitp4[3]));
+      }
+    }
+  }
+
+  if (!skipECAL_) {
+    auto const &ebRecHits = iEvent.get(ebRecHitsTag_);
+    inputs.reserve(inputs.size() + ebRecHits.size());
+    for (const auto &hit : ebRecHits) {
+      if (passedEcalNoiseCut(hit, thresholds)) {
+        const auto &hitp4 = getHitP4(hit.id(), hit.energy(), caloGeometry);
+        inputs.emplace_back(fastjet::PseudoJet(hitp4[0], hitp4[1], hitp4[2], hitp4[3]));
+      }
+    }
+
+    auto const &eeRecHits = iEvent.get(eeRecHitsTag_);
+    inputs.reserve(inputs.size() + eeRecHits.size());
+    for (const auto &hit : eeRecHits) {
+      if (passedEcalNoiseCut(hit, thresholds)) {
+        const auto &hitp4 = getHitP4(hit.id(), hit.energy(), caloGeometry);
+        inputs.emplace_back(fastjet::PseudoJet(hitp4[0], hitp4[1], hitp4[2], hitp4[3]));
+      }
+    }
+  }
+
+  bge_.set_particles(inputs);
+  iEvent.put(std::make_unique<double>(bge_.rho()));
+}
+
+std::array<double, 4> FixedGridRhoProducerFastjetFromRecHit::getHitP4(const DetId &detId,
+                                                                      double hitE,
+                                                                      const CaloGeometry &caloGeometry) const {
+  const CaloSubdetectorGeometry *subDetGeom = caloGeometry.getSubdetectorGeometry(detId);
+  const auto &gpPos = subDetGeom->getGeometry(detId)->repPos();
```

Comment:
I take it that these pointers are always guaranteed to be valid (at least, the way they are used in this plugin). Is that correct?

Reply:
yes I think so